pkg/log/zap/flags.go

```diff
@@ -60,26 +60,63 @@ func (ev *encoderFlag) Set(flagValue string) error {
 	val := strings.ToLower(flagValue)
 	switch val {
 	case "json":
-		ev.setFunc(newJSONEncoder())
+		ev.setFunc(newJSONEncoder)
 	case "console":
-		ev.setFunc(newConsoleEncoder())
+		ev.setFunc(newConsoleEncoder)
 	default:
 		return fmt.Errorf("invalid encoder value \"%s\"", flagValue)
 	}
 	ev.value = flagValue
 	return nil
 }
 
-func newJSONEncoder() zapcore.Encoder {
+func newJSONEncoder(ecfs []EncoderConfigOption) zapcore.Encoder {
 	encoderConfig := zap.NewProductionEncoderConfig()
+	for _, f := range ecfs {
+		f(&encoderConfig)
+	}
 	return zapcore.NewJSONEncoder(encoderConfig)
 }
 
-func newConsoleEncoder() zapcore.Encoder {
+func newConsoleEncoder(ecfs []EncoderConfigOption) zapcore.Encoder {
 	encoderConfig := zap.NewDevelopmentEncoderConfig()
+	for _, f := range ecfs {
+		f(&encoderConfig)
+	}
 	return zapcore.NewConsoleEncoder(encoderConfig)
 }
 
+type timeEncoderFlag struct {
+	setFunc func(zapcore.TimeEncoder)
+	value   string
+}
+
+var _ pflag.Value = &timeEncoderFlag{}
+
+func (ev *timeEncoderFlag) String() string {
+	return ev.value
+}
+
+func (ev *timeEncoderFlag) Type() string {
+	return "timeEncoder"
+}
+
+func (ev *timeEncoderFlag) Set(flagValue string) error {
+	var encoder zapcore.TimeEncoder
+	val := strings.ToLower(flagValue)
+	switch val {
+	case "iso8601", "rfc3339", "rfc3339nano", "millis", "nanos", "epoch":
+		if err := encoder.UnmarshalText([]byte(val)); err != nil {
+			return fmt.Errorf("unable to unmarshal, \"%s\"", val)
+		}
+	default:
+		return fmt.Errorf("invalid value for time encoder, \"%s\"", val)
+	}
```

Comment:
However, we would not be able to return error on `invalid` values explicitly, as the default is `EpochTimeEncoder` returned by `UnmarshalText`
https://github.com/uber-go/zap/blob/c9ca99951c74a6fcaf507962e772861e2218ce2e/zapcore/encoder.go#L174
I  believe that is how we wanted the design initially.  @hasbro17 WDYT?

Reply:
That makes sense. I'm surprised an unmarshaler doesn't return an error for an invalid value.

Reply:
True, But that still relies assuming `Unmarshaltext` would always return `epoch` as default in future as well. Either ways there is still dependency.

Reply:
I agree that it's weird that `UnmarshalText` never returns an error, but that seems to be an explicit choice of the zap maintainers. I guess my point of view is that we're basically plumbing the flag value through to the library and having it return what it thinks is the best thing to return.

So maybe it is reasonable to silently default to `epoch` when someone passes `--zap-time-encoder=foobar` because that lines up with what zap itself is doing.

Reply:
I'm fine with defaulting to whatever UnmarshalText gives us.
And maybe we can mention that invalid values will result in the default zap time encoder in the flag help text.